### PR TITLE
Gitleaks

### PR DESCRIPTION
**Pull request by urvashigupta7**
# Changes
This task makes it possible to use Gitleaks within Tekton pipelines.

# Submitter Checklist
- [x] Follows the [authoring recommendations](https://github.com/tektoncd/catalog/blob/main/recommendations.md)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

### CODE REVIEW

File: task/gitleaks/0.1/gitleaks/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+* **output_format** : output_format to use. (JSON|CSV|SARIF) (default: `json`)
+* **report_output_path** : path of file to save analysis report.
+* **args** : args. (default: `[]`)
+
+> _Note_ :  Gitleaks provides some default rulesets for detecting secrets which you can find [here](https://github.com/zricethezav/gitleaks/blob/master/config/default.go). However, user is free to create rulesets as per his requirement by defining them inside (.toml) format file. To know how to write these config file, you can visit this [link](https://github.com/zricethezav/gitleaks#rules-summary). Also, if you want to checkout an example config file, please head over to this [link](https://github.com/urvashigupta7/secret_detection/blob/master/gitleaks.toml). The config file can exist in the repository to be scanned or some other repository, if the config file is present in some other repository, you would need to provide url of **config_file_url** and the **config_file_path** where config file would be saved after fetching else you need to provide **config_file_path** only.
+
+
+## Usage
+
+```
```
Comment (vinamra28) [inline task/gitleaks/0.1/gitleaks/README.md:34]: nit:
```suggestion
```yaml
```

File: task/gitleaks/0.1/gitleaks/gitleaks.yaml [hunk 2]
```diff
@@ -0,0 +1,81 @@
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: gitleaks
+  labels:
+    app.kubernetes.io/version: "0.2"
```
Comment (vinamra28) [inline task/gitleaks/0.1/gitleaks/gitleaks.yaml:6]: nit
```suggestion
    app.kubernetes.io/version: "0.1"
```

File: task/gitleaks/0.1/gitleaks/gitleaks.yaml [hunk 3]
```diff
@@ -0,0 +1,81 @@
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: gitleaks
```
Comment (vinamra28) [inline task/gitleaks/0.1/gitleaks/gitleaks.yaml:4]: can you please keep this file in following path `task/gitleaks/0.1/gitleaks.yaml` ?

**Review by vinamra28 — CHANGES_REQUESTED** [review]
hi @urvashigupta7, thanks for your contribution :+1: 
Can you please follow the directory structure as below as the current one is violating the [TEP](https://github.com/tektoncd/community/blob/main/teps/0003-tekton-catalog-organization.md)?
```
task/
    gitleaks/
          0.1/
             gitleaks.yaml
             README.md
             samples/
             tests/
```

**Comment by urvashigupta7** [comment]
Hey @vinamra28 . After observing multiple task tests already there in the repo, I have tried to write the tests. Could you please check if they seem correct to you? Can we define multiple pipelines in **tests/run.yaml** ?

**Comment by urvashigupta7** [comment]
Hey @vinamra28. All tests passed. Had to create pvc for each pipeline, as git-clone task was giving error.

File: task/gitleaks/0.1/gitleaks.yaml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+         if [ -n "$(params.config_file_url)" ]
+         then
+          curl "$(params.config_file_url)" --output "$(params.config_file_path)"
+          echo "Fetched the config file from given ($(params.config_file_url)) URL and successfully saved at $(workspaces.source.path)/$(params.config_file_path)"
+         else
+          echo "No config file url was set"
+         fi
+
+    - name: detect-secret
+      image: zricethezav/gitleaks@sha256:55b2e1831c7395f17e9da3649924736b396b25be69efdc7a0137073fe0b0f456
```
Comment (vinamra28) [inline task/gitleaks/0.1/gitleaks.yaml:60]: can we use fully qualified name for the image?

File: task/gitleaks/0.1/gitleaks.yaml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      type: string
+      description: url from where the config file would be fetched.
+      default: ""
+    - name: config_file_path
+      type: string
+      description: path to config file.
+      default: ""
+    - name: output_format
+      type: string
+      description: output_format to use. (JSON|CSV|SARIF) (default:"json")
```
Comment (nikhil-thomas) [inline task/gitleaks/0.1/gitleaks.yaml:34]: this is a bit confusing.
shall we keep all the options and default in the same case

File: task/gitleaks/0.1/gitleaks.yaml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+      description: "args"
+      default: []
+  steps:
+    - name: fetch-config-file
+      image: registry.access.redhat.com/ubi8/ubi-minimal:8.2
+      workingDir: $(workspaces.source.path)
+      script: |
+         #!/usr/bin/env bash
+
+
```
Comment (nikhil-thomas) [inline task/gitleaks/0.1/gitleaks.yaml:50]:
```suggestion

```

**Review by vinamra28 — COMMENTED** [review]
can we improve the commit message so that it's bit more clear and then it can be merged?
/lgtm

File: task/gitleaks/0.1/samples/run.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+spec:
+  workspaces:
+  - name: workspace
+  tasks:
+  - name: clone-git-repo
+    workspaces:
+    - name: output
+      workspace: workspace
+    taskRef:
+      name: git-clone
```
Comment (PuneetPunamiya) [inline task/gitleaks/0.1/samples/run.yaml:15]: Should we use bundle here ??

Comment (PuneetPunamiya) [reply task/gitleaks/0.1/samples/run.yaml:15]: cc @afrittoli  @vdemeester  @piyush-garg

Comment (urvashigupta7) [reply task/gitleaks/0.1/samples/run.yaml:15]: Hey @PuneetPunamiya. I don't have much idea about bundle. Could you please elaborate why it should be used here?

Comment (PuneetPunamiya) [reply task/gitleaks/0.1/samples/run.yaml:15]: Using Tekton Bundle we can refer the task by creating a bundle of that task and pushing it to some OCI registry for e.g.

```
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: <pipeline-name>
spec:
  taskRef:
    name: <task-name>
    bundle: <image-name>   ---> Registery to which image for Task is pushed
```

But I think I'm wrong and we don't need this here

Comment (urvashigupta7) [reply task/gitleaks/0.1/samples/run.yaml:15]: Okay Thanks 😄

**Comment by urvashigupta7** [comment]
Hey @vinamra28. I have added small description about the tool and changed the 2nd commit message.

**Comment by vinamra28** [comment]
@urvashigupta7 probably can you squash both the commits into one so that it is cleaner?

**Comment by urvashigupta7** [comment]
Done @vinamra28

**Review by vinamra28 — APPROVED** [review]
/meow
Thanks :call_me_hand:

**Comment by chmouel** [comment]
nice work 👏🏻 

/lgtm
/approve

**Pull request merged**